### PR TITLE
Build compiler as a library

### PR DESCRIPTION
**Pull request by RazvanN7**
At the moment, I had to exclude ddmd/backend/outbuf.d from the library since the compilation would fail with the message : 

..\generated\windows\release\32\dmd.lib: Error: multiple definition of outbuf_829_3b2: ?reserve@Outbuffer@@QAEXI@Z and cgobj: ?reserve@Outbuffer@@QAEXI@Z
..\generated\windows\release\32\dmd.lib: Error: multiple definition of outbuf_829_3b2: ?clearn@Outbuffer@@QAEXI@Z and mscoffobj: ?clearn@Outbuffer@@QAEXI@Z
..\generated\windows\release\32\dmd.lib: Error: multiple definition of outbuf_829_3b2: ?writeByte@Outbuffer@@QAEXH@Z and outbuf: ?writeByte@Outbuffer@@QAEXH@Z
..\generated\windows\release\32\dmd.lib: Error: multiple definition of outbuf_829_3b2: ?writeShort@Outbuffer@@QAEXH@Z and cv8: ?writeShort@Outbuffer@@QAEXH@Z

In order to compile the library I had to exclude outbuf.d from the library and add it when the binary is built. I know this is not a solution, but I have no idea how to workaround this.

### CODE REVIEW

**Comment by mathias-lang-sociomantic** [comment]
Missing `git add src/ddmd/gencmain.d` maybe ? :)

**Comment by RazvanN7** [comment]
@mathias-lang-sociomantic yes, sorry... forgot about that

File: src/ddmd/mars.d [hunk 1]
```diff
@@ -186,67 +187,6 @@ Where:
 ", FileName.canonicalName(global.inifilename), fpic, m32mscoff, mscrtlib);
 }
 
-/// DMD-generated module `__entrypoint` where the C main resides
-extern (C++) __gshared Module entrypoint = null;
-/// Module in which the D main is
-extern (C++) __gshared Module rootHasMain = null;
```
Comment (mathias-lang-sociomantic) [inline src/ddmd/mars.d]: I'm sure @ibuclaw will be very happy about this one

File: src/ddmd/gencmain.d [hunk 2]
```diff
@@ -0,0 +1,71 @@
+module ddmd.gencmain;
```
Comment (mathias-lang-sociomantic) [inline src/ddmd/gencmain.d]: Module doc ? :)

File: src/ddmd/mars.d (hunk 1 above)
Comment (RazvanN7) [inline src/ddmd/mars.d]: I had to do it in order to exclude mars.d (where main is located) from the library

Comment (mathias-lang-sociomantic) [inline src/ddmd/mars.d]: I know. That was attempted 2 years ago but [didn't get pass Walter](https://github.com/dlang/dmd/pull/5282#r45561126).

**Comment by WalterBright** [comment]
The solution is to import outbuf.d, but do not compile (i.e. generate code) for it, as that would be redundant with outbuf.c. The function bodies that exist in outbuf.d are there for inlining purposes.

**Comment by RazvanN7** [comment]
@WalterBright I tried excluding all the files defined in "BACK_HDRS" from the library and then, when building the binary to introduce the directory with -I, but that doesn't work.

**Comment by WalterBright** [comment]
> that doesn't work.

Please be more specific.

File: src/posix.mak [hunk 3]
```diff
@@ -388,8 +388,11 @@ ifdef ENABLE_LTO
 $G/dmd: $(DMD_SRCS) $(ROOT_SRCS) $G/newdelete.o $G/lexer.a $(G_GLUE_OBJS) $(G_OBJS) $(STRING_IMPORT_FILES) $(HOST_DMD_PATH)
 	CC=$(HOST_CXX) $(HOST_DMD_RUN) -of$@ $(MODEL_FLAG) -vtls -J$G -J../res -L-lstdc++ $(DFLAGS) $(filter-out $(STRING_IMPORT_FILES) $(HOST_DMD_PATH),$^)
 else
-$G/dmd: $(DMD_SRCS) $(ROOT_SRCS) $G/newdelete.o $G/backend.a $G/lexer.a $(STRING_IMPORT_FILES) $(HOST_DMD_PATH)
-	CC=$(HOST_CXX) $(HOST_DMD_RUN) -of$@ $(MODEL_FLAG) -vtls -J$G -J../res -L-lstdc++ $(DFLAGS) $(filter-out $(STRING_IMPORT_FILES) $(HOST_DMD_PATH) $(LEXER_ROOT),$^)
+$G/dmd: $G/dmd.a $D/mars.d $(STRING_IMPORT_FILES)
+	CC=$(HOST_CXX) $(HOST_DMD_RUN) -of$@ $(MODEL_FLAG) -I$D -L-lstdc++ -J$G -J../res $(DFLAGS) $G/dmd.a $D/mars.d
+
+$G/dmd.a: $(DMD_SRCS) $(ROOT_SRCS) $G/newdelete.o $G/backend.a $G/lexer.a $(STRING_IMPORT_FILES) $(HOST_DMD_PATH)
+	CC=$(HOST_CXX) $(HOST_DMD_RUN) -lib -of$@ $(MODEL_FLAG) -vtls -J$G -J../res -L-lstdc++ $(DFLAGS) $(filter-out $(STRING_IMPORT_FILES) $(HOST_DMD_PATH) $(LEXER_ROOT) $D/mars.d,$^)
```
Comment (WalterBright) [inline src/posix.mak]: Here you're using the compiler to build the library, but in the failing autotester build it's using `ar` and `ranlib`. Use the compiler.

File: src/ddmd/gencmain.d (hunk 2 above)
Comment (ibuclaw) [inline src/ddmd/gencmain.d]: Significant contributions need a licence included at the top.

File: src/ddmd/mars.d (hunk 1 above)
Comment (ibuclaw) [inline src/ddmd/mars.d]: I'll be happy when `grep 'import ddmd.mars;' ` returns no results, as that would unblock one of the stoppers for gdc/ddmd.  If this does just that then please link in the bugzilla report I created.

Comment (RazvanN7) [inline src/ddmd/mars.d]: This does more than that, but if the PR passes I will close the bug report manually and link the PR

**Comment by RazvanN7** [comment]
@braddr any ideas how we can get past the failing autotester test?

File: src/posix.mak (hunk 3 above)
Comment (RazvanN7) [inline src/posix.mak]: I am also using `ar` and `ranlib` to create `backend.a`. From what I've seen, the autotester uses the compiler to build `dmd.a` and `ar` for `backend.a`. Could you please be more explicit?

**Comment by jacob-carlborg** [comment]
Does this PR make sense when we have parts of the frontend available as a DUB package?

**Pull request closed without merging**